### ISSUE TITLE
Sync Annual Nth recurrences rules as 12 monthly O->G

### ISSUE THREAD

**Issue by HCITMK** [phw198/OutlookGoogleCalendarSync#445]
:memo: Use the below template to report your issue.  
:bulb: Have you upgraded to the latest alpha to check if the problem is already fixed?  
:paperclip: To aid investigation, please provide your OGcalsync.log file ([see instructions](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Reporting-problems#posting-your-ogcalsynclog-file)).  

----

**OGCS Version**: v1.2.3.4  
**Installed or Portable**: Installed

**Problem Description**  
\<Update here>An appointment with a recurrence, say 1st Thursday in May every year, is not corrently reflected in gmail.

**Steps to Reproduce the Issue**  
1. \<Update>
1. 

**Any other information**  
\<Update here>

**Comment by phw198** [comment]
You've not provided the version of OGCS you are running.

**Comment by HCITMK** [comment]
My apologies, I am new to OGCS.  I previously used Sony Companion quite successfully on Windows XP, but it won't run on Windows 10, so I'm looking around for an alternative application and came across yours.

I am running version 2.6.5.0

I have just set up two repeating events for today , 'Repeat Yearly' and 'Repeat 12 Monthly', both with the  'fourth Friday of January' recurrence and run OGCS.  Both events appear in Google Calendar as Repeat 'Annualy on 26 January'?  And on 26 January 2019, not the 4th Friday.  A log is attached.

[OGcalsync.log](https://github.com/phw198/OutlookGoogleCalendarSync/files/1667084/OGcalsync.log)

**Comment by HCITMK** [comment]
I've downloaded setup.exe for version 2.6.6.0, but when run, Windows 10 intervenes with: "Windows Defender SmartScreen prevented an unreconised app from starting. Running this app might put your PC at risk" , without an override option?

**Comment by phw198** [comment]
OK, thanks I will investigate the annual recurrence issue.

For the v2.6.6.0 download, check if Edge has unhelpfully blocked the downloaded file - right click, Properties and click `Unblock`? Alternatively, in OGCS > About > Check for Alpha updates >Check for Update.

**Comment by phw198** [comment]
No, I can't replicate this. I created a recurring annual event on 4th Friday of January (26th for 2018), and OGCS replicates this as
```
2018-01-27 11:24:29,265 DEBUG  11 OutlookGoogleCalendarSync.GoogleOgcs.Calendar [317] -  Processing >> 26/01/2018 15:00 (R) => "TEST"
2018-01-27 11:24:29,306 DEBUG  11 OutlookGoogleCalendarSync.Recurrence [31] -  Creating Google iCalendar definition for recurring event.
2018-01-27 11:24:29,386 DEBUG  11 OutlookGoogleCalendarSync.Recurrence [40] -  RRULE:FREQ=YEARLY;BYDAY=FR;BYMONTH=1;BYSETPOS=4
```
Checking in Google, the new interface doesn't know how to show it as it's too "complicated", but the meeting shows on Friday 25th Jan 2019 and Friday 24th Jan 2020.

**Comment by HCITMK** [comment]
Paul,

Neither can I now!  Perhaps this is because I seem to be using a different version of OGCS?  Yesterday, I was running what was reported as version 2.6.5.0 whereas today it is reported as version 2.6.0.0?  Have I managed to get the Alpha version running, or have I reverted to an older version?

Regards,

Andy Dransfield

**Comment by phw198** [comment]
v2.6.5 would be the more recent alpha and v2.6.0 the older beta. I would expect the problem to be in the beta if at all. 

FYI new beta v2.7 is just out.

**Comment by HCITMK** [comment]
Paul,

I’ve now implemented v2.7 reasonably successfully, but I’m afraid that I commented too quickly.  I want to synchronise Outlook with Google Calendar and then use an app on my smart Android phone to monitor my calendar in near-real time.  Unfortunately, none of the apps handle annual repeats properly and give some strange manifestations.  One app repeats an annual event by repeating it on the selected day of each week in a defined month rather than the nth week.  Another app translates the annual repeat on a defined nth weekday of a specified month as on the nth day of the month and so moves on the day in subsequent years?!?

The workaround for these apps is to define the annual repeat as a 12 monthly repeat.  Unfortunately, if I set up a 12 monthly repeat in Outlook, OGCS converts that into an annual repeat in Google.  Fortunately, if I then change the event from Annual to 12 monthly in Google, OGCS respects the change until I make a change in Outlook after which OGCS changes the event in Google back to an annual repeat.  Can OGCS respect the difference between an annual and a 12 monthly repeat in the next release?

Regards,
Andy Dransfield

**Comment by phw198** [comment]
> Can OGCS respect the difference between an annual and a 12 monthly repeat in the next release?

I see what you mean, Google also allows for the recurrence in this way:
![image](https://user-images.githubusercontent.com/11843015/35527456-aeafe9ba-0522-11e8-8896-f7e518b1b044.png)

Though the root bug seems to be with other Android apps not being able to interpret the iCal standard properly for an annual recurrence. However, I've checked my previous test on the [Google Android calendar app](https://play.google.com/store/apps/details?id=com.google.android.calendar) and it displays the events fine. Again, if I go to edit the appointment on my phone it shows it as "annual recurrence", but it's actually shows the events on the nth Friday of the month (so the same thing the web interface struggles with). Not a problem unless you want to edit the recurring pattern in Google though.

**Comment by phw198** [comment]
I have tested an Outlook monthly recurrence with an interval of 12 and unfortunately the API surfaces it as an annual recurrence - it's being too "clever" for it's own good. I assume it is actually held differently deeper in the Outlook object model as Outlook remembers the difference if I save it either as an annual recurrence or as every 12 months.

So back to your question "can OGCS respect the difference" - no it can't because Microsoft doesn't! :roll_eyes:

**Comment by HCITMK** [comment]
Paul,

Very intriguing, so I reverted to my old synchronise program, Sony PC companion, on Windows XP using the same Outlook files.  None of the apps that I have on my Android smart phone can create events that repeat annually on, say, the 1st Monday of February, but they can all create 12 monthly repeat events for that series of Mondays.  Not surprisingly, the events all synchronise correctly into Google Calendar as either the 5th February annually or 1st Monday of February 12 monthly.   Google Calendar cannot create a repeated annual event either for the 1st Monday of February.  So far as expected!  I also set up annually repeating and 12 monthly repeating events in Outlook for the 1st Monday in February.

Now the differences.  After running Sony PC Companion, both Outlook events synchronised with Google Calendar as 12 monthly events on 1st Monday of February.  This is as I would expect as Google Calendar doesn’t have the annually repeat function for the 1st Monday of February.  Similarly, both of my Android events synchronised with Outlook exactly as expected.

So, I accept that OGCS cannot respect the Outlook differences between annual and 12 monthly events, but it could synchronise an annual event on the 1st Monday of February as a 12 monthly event on the 1st Monday of February rather than, incorrectly, as an annual event on 5th February.  I appreciate that the API sees both as annual recurrences, but you could then write them to Google Calendar as 12 monthly events.  You should be able to see them both as the 1st Monday because Sony PC Companion can, though on a different Windows platform.

Quite separately, having used Sony PC Companion, I now experience the OGCS errors that I encountered when I first tried OGCS, namely the error message:

Appointment update failed
Timezone “WET” has no mapping.
They are all associated with my 12 monthly repeating events.

Regards,
Andy Dransfield

**Comment by phw198** [comment]
So let's be clear, OGCS is not doing anything "incorrectly". 

If you use https://www.textmagic.com/free-tools/rrule-generator and plug in our example, it produces an RRULE that is exactly the same as what you see in the OGCS logs: `FREQ=YEARLY;BYDAY=FR;BYSETPOS=4;BYMONTH=1`

![image](https://user-images.githubusercontent.com/11843015/35572335-4c3b9e28-05cc-11e8-9ec7-65c515fdd2ac.png)

What you are asking for is an enhancement to make it more compatible with the shortcomings of other apps (indeed even the Google Calendar web app) :slightly_smiling_face: 

So we would need to say, if Outlook has `olRecurringYearlyNth` recurrence with an interval of 12, translate this to `FREQ=MONTHLY;BYDAY=FR;BYSETPOS=4;INTERVAL=12`.

And, strictly, this would need to be done for two-way sync back in to Outlook too, else we would trample the original yearly recurrence with a Google-pleasing 12 monthly recurrence. Do-able I guess...

Re the [WET time"zone" that isn't](https://en.wikipedia.org/wiki/Western_European_Time) - you've contaminated your appointments with a time **region** of UTC+1 which has no mapping to a unique [IANA timezone](https://en.wikipedia.org/wiki/Tz_database). You can see support for [CET is deprecated](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones) from IANA. 

On a system that's not XP :wink: you'll need to edit each problematic appointment and pick a valid timezone.

**Comment by HCITMK** [comment]
Paul,

I apologise if I have inadvertently annoyed you – I really am trying to be helpful!

My “incorrect” comment is about the translation by OGCS from the 1st Monday in February annually to the 5th of February annually, which may be the same this year, but not next.  Surely, it is better to change the type of repeat rather than introduce an unintended change?  Doesn’t OGCS currently convert the 1st Monday of February 12 monthly into 5th February annually?!?

As for the other direction, I could see what Sony PC Companion does, but, in my view, it is more defensible to change from yearly to 12 monthly.  Anyway, currently, OGCS will ‘convert’ 1st Monday of February in Outlook to 5th February in Google Calendar.  What does it do in the other direction?  Does it convert 5th February back to the 1st Monday, or leave it as the 5th February?  Which is the bigger ‘trample’?

Where do I need to edit the timezones, Outlook or Google Calendar?  I’m assuming that the problem was introduced by PC Sony Companion, so I don’t expect it to appear again once I’ve fixed the issue.

Regards,
Andy Dransfield

**Comment by phw198** [comment]
@HCITMK No worries - we'll get the bottom of this!

> the translation by OGCS from the 1st Monday in February annually to the 5th of February annually, which may be the same this year, but not next.  Surely, it is better to change the type of repeat rather than introduce an unintended change?

The point is that OGCS is **not** doing that and not introducing any change. It is following the iCAL standard and defines a rule for Google that is `RRULE:FREQ=YEARLY;BYDAY=MO;BYMONTH=2;BYSETPOS=1`. See, nowhere does that say "repeat on the 5th February". You need to check what is actually happening on the Google calendar as close to the truth as you can get: https://calendar.google.com - and this displays as you want. If you want, you could export the Google calendar as `.ics` and confirm the appointment is still defined with that RRULE.

:thought_balloon: If you have an Android app that can't handle this recurrence, then it may be possible this is corrupting it to something else (ie repeat on 5 Feb each year) and by the time you check Google, it's wrong again and you're concluding OGCS is doing this?

What app are you using that can't handle this recurrence? If it's free I could try replicating the issue. Have you tried the official Google calendar app?

I do take your point however, that a) you can't view or edit the annual recurrence rule in Google; b) other dumb apps won't even display the annual recurrence series on the right days at all (taking your word on this). So for those reasons, I'm willing to entertain a workaround enhancement (not bugfix!) to OGCS.

> Where do I need to edit the timezones, Outlook or Google Calendar?

It sounds like Outlook, but in whichever is the source calendar for the particular direction of sync when the error occurs. If you're one-way syncing O -> G, then edit in Outlook.

**Comment by phw198** [comment]
So I've had a play with translating an annual recurrence to 12 monthly recurrence - have a play and see if the Google apps are happier.

:package: Development build [v2.7.0.1.zip](https://github.com/phw198/OutlookGoogleCalendarSync/files/1683604/v2.7.0.1.zip) ([see instructions for applying](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Reporting-Problems#applying-a-hotfix)).

Please let me know how it goes.

**Comment by HCITMK** [comment]
Paul,

 

Spot on!!!

 

On reflection, the real problem lies with Google Calendar that doesn’t implement the nth weekday of a month annually, and thus the Android apps can’t either.  Your latest release recognises that and is an effective solution to the Google problem.  All three of my Android apps now reflect my Outlook calendar!

 

Regards, and thanks,

Andy Dransfield

**Comment by HCITMK** [comment]
Paul,

Spot on!!!

On reflection, the real problem lies with Google Calendar that doesn’t implement the nth weekday of a month annually, and thus the Android apps can’t either.  Your latest release recognises that and is an effective solution to the Google problem.  All three of my Android apps now reflect my Outlook calendar!

Regards, and thanks,
Andy Dransfield

P.S. Now for the Contacts version – when is that likely to be available?  APD

**Comment by phw198** [comment]
Yes exactly - and the purest in me hates this workaround :disappointed: 

I may further enhance this as a configurable option for the user, though.

As OGCS is a spare time project for me, I can't commit to dates for new features like contacts....but I'm coming round to the idea (upvote [here](https://github.com/phw198/OutlookGoogleCalendarSync/issues/407)) as the syncing of calendars is becoming more stable. 

You can [check here](https://github.com/phw198/OutlookGoogleCalendarSync/projects/3) to see what's coming down the pipeline - calendar colour syncing and multiple calendar support first.

**Comment by HCITMK** [comment]
Paul,

As you may have realised, I also am a software developer, Open Database Connectivity Drivers (ODBC), so I share your purist views and would be content with an option about how to treat 4th Tuesday of February annual repeat as there is no pure way to convert that into Google Calendar.  Personally, I have set up all my annual events as repeating 12 monthly.

However, I risk upsetting you again by saying that converting 12 monthly repeats into annual repeats was a bug and I hope that you don’t apply your option to them.

I have added to the long list (one) of requests for a Contacts program.

Andy

**Comment by phw198** [comment]
>converting 12 monthly repeats into annual repeats was a bug

A bug in the sense that it's a limitation of the Outlook API - so the bug is Microsoft's not mine ;-)

So the options for "nth weekday of month recurrences" are:
1. Native/Purist (v2.7.0.0) = rule doesn't display/isn't editable properly in Google apps
   1. Syncs annual recurrence as is to Google 
   1. Syncs 12 monthly recurrence to Google as annual recurrence (Outlook API bug)
1. Workaround (v2.7.0.1) = rule editable in Google apps
   1. Syncs annual recurrence to Google as 12 monthly recurrence
   1. Sync 12 monthly recurrence as 12 monthly recurrence (translate Outlook API bug)

As it's not possible to cherry pick 1i and 2ii I think you're right, the workaround is the lesser of the two evils and overall plays nicer with Google so it's probably not worth surfacing this as a choice to the user; I'll bake it into the next release as v2.7.0.1 is.

Thanks for your help on this and the donation! :+1:

**Comment by HCITMK** [comment]
Paul,

Any progress on your Outlook Google Contacts Sync App?

Andy

**Comment by phw198** [comment]
Hi Andy, 
Thanks for checking, but no not yet. Quite some growth pains the code needs to go through first in preparation, but I do now want to get Contacts sync in to OGCS.

**Comment by HCITMK** [comment]
Paul,

 

I’m now getting the following when trying to run Outlook Google Calendar Sync:

 

Could not load file or assembly ‘office, Version=12.0.0.0,

Culture=neutral, PublicKeyToken=71e9bce111e9429c’ or one

Of its dependencies. The system cannot find the file specified.

 

I rather suspect a Microsoft update that went through a few days ago.  Before reloading Microsoft Office, I thought that I’d check with you first.  Has this issue been reported to you?

 

Regards,

Andy Dransfield



---
This email has been checked for viruses by AVG.
https://www.avg.com

**Comment by phw198** [comment]
Hi @HCITMK - Looks like issue #455 which was fixed in v2.7.1.

If you are on an earlier version, due to a [GitHub change](https://github.com/phw198/OutlookGoogleCalendarSync/issues/497) you will need to [download the latest version directly from the project website](https://phw198.github.io/OutlookGoogleCalendarSync/).

**Comment by HCITMK** [comment]
Paul,

Any progress on your Outlook Google Contacts Sync App?

I’ve just realised that Google Calendar Sync has not run for a few days, and on investigation I found that AVG had assigned it to Quarantine.  Are you aware?

Andy

**Comment by phw198** [comment]
Not started on Contacts sync yet, still a way off I'm afraid.

I'm aware that AVG is a medicine worse than the disease, yes :roll_eyes:   Seriously though, what version of OGCS are you running? I started code signing OGCS as of v2.7.3 so I would really hope that helps the likes of AVG from black listing it...

**Comment by HCITMK** [comment]
I was invited to install a new beta this morning which I duly did followed shortly afterwards by an AVG warning:

Threat Blocked

We've blocked OutlookGoogleCalendarSync.exe because it was infected with IDP.ALEXA.51.

What would you like to do?

I've temporarily move it to Quarantine.

Is this a known issue?

**Comment by phw198** [comment]
@HCITMK Unfortunately, I discovered the certificate authority I used do not provide code signing certificate renewals, so I had to purchase a new one. This has clearly lost the reputation the old certificate had built up, which is more than a little frustrating. Next year I will consider changing certificate authority for one that provides renewals so this does happen again (well, it will again next year, but that would then be the last time!)...

Note, that new certificate was only applied to v2.7.9 if you are installing a previous version of OGCS, then AVG is just being extra dumb.
